rules/falco_rules.yaml

```diff
@@ -2604,7 +2604,25 @@
   output: "Docker or kubernetes client executed in container (user=%user.name %container.info parent=%proc.pname cmdline=%proc.cmdline)"
   priority: WARNING
   tags: [container, mitre_execution]
+
+
+# This rule is not enabled by default, as there are legitimate use
+# cases for raw packet. If you want to enable it, modify the
+# following macro.
+- macro: consider_raw_packet_creation
+  condition: (never_true)
+
+- list: raw_packet_binaries
+  items: []
+
+- rule: Raw packet was created in a container
+  desc: Detect new raw packets in a container. raw packets could be used to do ARP Spoofing by attacker.
+  condition: evt.type=socket and evt.arg[0]=AF_PACKET and container and not proc.name in (raw_packet_binaries)
```

Comment:
1. can you rename the rule to `Raw socket created in container`
2. can you add another constraint to `evt.arg[1]=SOCK_RAW`? I haven't verified my self, but I would like to have the socket type as part of the condition.

Reply:
Hi, @Kaizhe 
Thank you for your review.

> 2. can you add another constraint to evt.arg[1]=SOCK_RAW?
> I haven't verified my self, but I would like to have the socket type as part of the condition.
- `AF_PACKET` is used for raw packet (layer 2). and socket_type could have `SOCK_RAW` and `SOCK_DGRAM` in my understanding.
  - This rule aims to catch a raw packet (Layer 2) connection to detect attacks like ARP Spoofing. So I couldn't add another constraint.
  - Ref: https://linux.die.net/man/7/packet
  - Ref: https://linux.die.net/man/7/raw

> 1. can you rename the rule to Raw socket created in container
- On Man page of packet(7), `raw packets` is used for AF_PACKET. so I used `raw packets`. what do you think about the naming?
  - And I agree `raw packets` is not clear well. So how about changing desc to this?
```
Detect new raw packets at the device driver (OSI Layer 2) level
 in a container. raw packets could be used to do ARP Spoofing by attacker.
```
  - Ref: https://linux.die.net/man/7/packet

Reply:
this is socket creation event (not receive or send), and I am not sure whether the ARP spoofing can be done via both `SOCK_RAW` and `SOCK_DGRAM`. I think it can be done with `SOCK_RAW` (https://opensourceforu.com/2015/03/a-guide-to-using-raw-sockets/) and further more, `PF_PACKET` seems similar to `AF_PACKET`. This quite depends on the underneath system.

Regarding the naming, I would suggest `raw socket created in container` if the `raw` type is the one must be in the condition.

Reply:
Thank you for your comment!

Yes, I should add PF_PACKET too! I'll fix it.

About `SOCK_DGRAM`, I found `SOCK_DGRAM` is used in some tools for ARP Spoofing.
(`SOCK_RAW` is popular for ARP Spoofing though, because `sockaddr_ll` struct is required to use `SOCK_DGRAM`)
Ref: https://github.com/smikims/arpspoof/blob/244d9eedd0dc9576b062bc01963d878560d574f5/arpspoof.c#L319

And naming, I think "raw socket" is a more common word. so I'll rename "raw packet" to "raw socket". and I'll add the explanation to desc

Reply:
@Kaizhe How about `packet socket created in container` for rule name?
When using `raw socket`, I think user could misunderstand L2 Layer as L3 Layer.
(i.e. `ping` uses raw socket(L3 Layer), but rule detects packet socket(L2 Layer), so user doesn't have to add `ping` to `raw_packet_binaries`. )
So I felt `packet socket` is more understandable.
Ref: https://linux.die.net/man/7/packet

Reply:
I sent new commit as you suggested. Thank you!
  - I renamed macro and list.
  - I fixed condition.
    - (added `PF_PACKET` and `consider_raw_packet_creation` macro).

- This is not resolved yet
   - Naming problem about `raw packet`.
   - Whether adding `SOCK_RAW` to condition.

Reply:
- I did investigation about socket_family. As far as I tested, Falco needs `AF_PACKET` only (`PF_PACKET` is not needed in condition).
  - because both are same value when calling syscall. and it seems both value is handled as `AF_PACKET`

### Investigation
#### PF_PACKET value on glibc
- AF_PACKET and PF_PACKET is same value.(17 = 0x11)
```
#define AF_PACKET	PF_PACKET
```
- Ref: https://sources.debian.org/src/glibc/2.29-3/sysdeps/unix/sysv/linux/bits/socket.h/

#### Confirmation when syscall
- both socket uses same value(0x11) for fist argument when syscall.
  - (Falco detected both socket() using AF_PACKET)
```
$ cat socktest.c 
#include <sys/socket.h>
#include <linux/if_packet.h>
#include <net/ethernet.h>
#include <arpa/inet.h>

int main() {
    int fd = socket(AF_PACKET, SOCK_RAW, htons(ETH_P_ARP));
    fd = socket(PF_PACKET, SOCK_RAW, htons(ETH_P_ARP));
}

$ gcc socktest.c -o socktest
$ sudo strace -e trace=socket -e raw=socket ./socktest
socket(0x11, 0x3, 0x608)                = 0x3
socket(0x11, 0x3, 0x608)                = 0x4
+++ exited with 0 +++
$ 
```

Reply:
@rung thanks for the further investigation! Really appreciate it!